### ISSUE TITLE
[Parity 9/16] Serialization and interop - safetensors, HuggingFace, TensorBoard, .pt/pickle, sharded checkpoints

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#218]
Part of epic #209.

## Summary

Users expect to load HuggingFace checkpoints, safetensors files, PyTorch `.pt` files, and to write TensorBoard logs. We have GGUF (unique!) and a custom serializer in `Engines/Compilation/Serialization/`, but the common-interchange formats are missing.

### Core serialization
- `torch.save(obj, f, pickle_protocol)` / `torch.load(f, map_location)`
- `state_dict()` / `load_state_dict()` on modules, with strict/non-strict loading
- `torch.package` (self-contained model archive)

### Safetensors (ecosystem standard)
- Read & write, lazy-loading, zero-copy mmap, slicing without full load, metadata dict
- Used by HuggingFace Transformers, Diffusers, vLLM

### HuggingFace interop
- `AutoModel.from_pretrained` conventions require matching `config.json` + `model.safetensors` (or sharded `model-00001-of-000N.safetensors` + `model.safetensors.index.json`)

### TensorBoard
- `torch.utils.tensorboard.SummaryWriter` — scalars, images, histograms, embeddings, graphs, hyperparams

### Sharded / distributed checkpoints
- `torch.distributed.checkpoint` — save/load across FSDP ranks, consolidate, reshard

## Current state in the repo

- `src/AiDotNet.Tensors/NumericOperations/GgufReader.cs` — ✅ unique
- `src/AiDotNet.Tensors/Engines/Compilation/Serialization/` — custom format (inspect before integrating)
- No safetensors, no HF state_dict bridge, no TensorBoard, no `.pt` reader.

## How we beat PyTorch

1. **Safetensors with zero-copy `MemoryMappedFile` + `TensorStorage` aliasing**: PyTorch has zero-copy read but not write; we do both. Bonus: we can *append* tensors to a safetensors file in-place (common in fine-tune checkpoints) — PyTorch rewrites the whole file.
2. **`.pt` reader without Python runtime**: we parse PyTorch's pickle protocol enough to recover tensors, skipping Python code objects. HF `transformers` users get first-class C# loading.
3. **HF state_dict convention mapper**: a registry that knows the naming mismatches between HF naming (`bert.encoder.layer.0.attention.self.query.weight`) and our layer names. Ship presets for BERT, GPT-2, Llama, ViT, CLIP, T5, Stable Diffusion. PyTorch users write this by hand.
4. **Consolidated + sharded safetensors writer** with automatic shard-size targeting — PyTorch requires the `accelerate` library.
5. **TensorBoard writer with zero buffering overhead**: writes protobuf directly via `System.IO.Pipelines`, no string formatting. Target faster than `tensorboardX` on scalar-flood loads.
6. **Cross-format auto-convert CLI**: `dotnet tool` that converts `.pt` → `.safetensors` → `.gguf` without Python. PyTorch has no such tool.
7. **Quantized safetensors**: our int1/2/3/4/NF4/FP4/FP8 tensors serialize inline with a dtype tag that HF Transformers can read back once we contribute the dtype upstream. Even before that happens, we read our own format losslessly.
8. **Sharded distributed checkpoints** that integrate with FSDP (see distributed issue) — save per-shard, consolidate on read, resharding supported.

### Safetensors
- [ ] `SafetensorsReader` (mmap, zero-copy slices, metadata)
- [ ] `SafetensorsWriter` (consolidated + sharded)
- [ ] Sub-byte / FP8 dtype tags
- [ ] Appendable writer for fine-tune scenarios

### `.pt` / pickle
- [ ] `PtReader` — parse PyTorch pickle protocol 2/3, recover `torch.Tensor` → `Tensor<T>`
- [ ] `PtWriter` (optional — less important than reader)

### HuggingFace bridge
- [ ] `HFStateDictMapper` with convention registry
- [ ] Presets: BERT, GPT-2, GPT-Neo, Llama (v1/2/3), ViT, CLIP, T5, Whisper, Stable Diffusion UNet, SDXL
- [ ] `config.json` parser

### TensorBoard
- [ ] `SummaryWriter` — scalar, image, histogram, embedding, graph, hparams
- [ ] Protobuf-direct writer (no external dep)

### Sharded / distributed checkpoints
- [ ] `DistributedCheckpoint.Save/Load` — per-shard + consolidation
- [ ] Resharding on load (N shards → M shards)

### CLI tool
- [ ] `dotnet ai-tensors convert --from pt --to safetensors`
- [ ] `dotnet ai-tensors convert --from safetensors --to gguf --quant Q4_K`

## Acceptance criteria

- Round-trip tests: write → read → compare for every format.
- Load public HF model (tiny-llama / TinyBERT / ViT-Tiny) from `model.safetensors` and reproduce a reference logit tensor.
- Sharded safetensors: 10 GB model written in N shards, loaded, tensors match.
- `.pt` reader loads a PyTorch-saved state dict end-to-end.
- TensorBoard output opens in the real TensorBoard UI with scalars / images / histograms visible.
- FSDP sharded checkpoint: save across 4 ranks, reshard to 2 ranks, weights match.
- Benchmark: safetensors load throughput vs `safetensors` crate / PyTorch extension; we target within 10% on Linux, clear win on Windows (no mmap perf cliff).

## Implementation notes

- `src/AiDotNet.Tensors/Serialization/Safetensors/`
- `src/AiDotNet.Tensors/Serialization/Pt/`
- `src/AiDotNet.Tensors/Serialization/HuggingFace/`
- `src/AiDotNet.Tensors/Serialization/TensorBoard/`
- `src/AiDotNet.Tensors/Serialization/DistributedCheckpoint/`
- Reuse `NativeMemoryManager.cs` for mmap-backed storage aliasing.
- Reuse `TensorStorage.cs` so a safetensors view becomes a regular `Tensor<T>` with no copy.
- Pickle parser: ship a minimal reader that recognizes `torch.storage._TypedStorage`, `torch.Tensor.__setstate__`, and the common `OrderedDict` envelope.

**Comment by ooples** [comment]
# Implementation plan — detailed

## 1. Current state

| File | Status |
|---|---|
| `src/AiDotNet.Tensors/NumericOperations/GgufReader.cs` | ✅ Unique to us |
| `src/AiDotNet.Tensors/Engines/Compilation/Serialization/` | Custom format (audit coverage) |

No safetensors, no HF state_dict bridge, no TensorBoard, no `.pt` reader.

## 2. Scope — full inventory

### Safetensors
```
SafetensorsReader(stream, mmap=true) — zero-copy slices, metadata dict
SafetensorsWriter(stream) — consolidated + sharded modes
  - sub-byte / FP8 dtype tags
  - appendable writer for fine-tune (PyTorch rewrites whole file)
```

### .pt / pickle
```
PtReader(stream) — parse PyTorch pickle protocol 2/3, recover torch.Tensor → Tensor<T>
  - recognizes torch.storage._TypedStorage, torch.Tensor.__setstate__, OrderedDict
PtWriter (optional)
```

### HuggingFace bridge
```
HFStateDictMapper with convention registry
Presets: BERT, GPT-2, GPT-Neo, Llama (v1/2/3), ViT, CLIP, T5, Whisper,
         StableDiffusion UNet, SDXL
config.json parser
```

### TensorBoard
```
SummaryWriter(log_dir) — protobuf-direct via System.IO.Pipelines
  - AddScalar, AddScalars
  - AddImage, AddImages
  - AddHistogram
  - AddEmbedding
  - AddGraph (from LazyTensorScope)
  - AddHparams
```

### Sharded / distributed checkpoints
```
DistributedCheckpoint.Save(model, path, sharded=true)
DistributedCheckpoint.Load(model, path)
  - Resharding on load (N shards → M shards)
  - FSDP integration via #215
```

### CLI tool
```
dotnet ai-tensors convert --from pt --to safetensors
dotnet ai-tensors convert --from safetensors --to gguf --quant Q4_K
dotnet ai-tensors convert --from onnx --to gguf   # uses #169
```

**Total: ~25 surfaces.**

## 3. Win vectors

1. **Safetensors zero-copy read AND write** via `MemoryMappedFile` + `TensorStorage` aliasing. PyTorch has zero-copy read only; write re-serializes.
2. **Appendable safetensors** — add tensors to existing file in-place for fine-tune. PyTorch rewrites whole file.
3. **`.pt` reader without Python runtime** — minimal pickle parser recovers tensors. HF Transformers users get C# loading.
4. **HF state_dict convention registry** — presets for 10+ model families. PyTorch users write by hand.
5. **Consolidated + sharded safetensors writer** with auto shard-size targeting. PyTorch needs `accelerate`.
6. **TensorBoard protobuf-direct** via `System.IO.Pipelines` — faster than `tensorboardX`.
7. **Cross-format CLI** `.pt` → `.safetensors` → `.gguf` without Python. No PyTorch equivalent.
8. **Quantized safetensors** with dtype tags for int1/2/3/4/NF4/FP4/FP8. We read our own; once HF adopts, bidirectional.
9. **Sharded distributed checkpoints** integrate with #215 FSDP — resharding supported.

## 4. Per-axis beat targets

| Axis | Target |
|---|---|
| Safetensors load throughput vs Rust `safetensors` | within 10% Linux, clear win on Windows (no mmap cliff) |
| Safetensors write with appending | works; feature unique |
| `.pt` load tiny-BERT/tiny-Llama | full state dict recovered |
| HF model load (tiny-BERT/ViT-Tiny) | reproduce reference logits within 1e-4 |
| TensorBoard scalar-flood throughput | ≥ tensorboardX |
| Sharded 10GB safetensors roundtrip | works; reshard N→M validated |
| FSDP sharded save/load across 4→2 ranks | weights match |
| CLI convert pt→safetensors→gguf | end-to-end test |

## 5. File inventory

**New (~50):**
- `src/AiDotNet.Tensors/Serialization/Safetensors/` — SafetensorsReader, SafetensorsWriter, SafetensorsShardWriter, AppendableWriter, DtypeTag (~8)
- `src/AiDotNet.Tensors/Serialization/Pt/` — PtReader, PickleParser, PtWriter (~6)
- `src/AiDotNet.Tensors/Serialization/HuggingFace/` — HFStateDictMapper + 10 preset files (~12)
- `src/AiDotNet.Tensors/Serialization/TensorBoard/` — SummaryWriter, ProtoWriter, EventFile (~8)
- `src/AiDotNet.Tensors/Serialization/DistributedCheckpoint/` — Save, Load, Reshard (~6)
- `tools/AiDotNet.Tensors.Cli/` — new CLI project with `convert` subcommands (~10)

**Modified (~8):** `NativeMemoryManager.cs` (mmap alias), `TensorStorage.cs` (alias from mmap), `GgufReader.cs` (bidirectional), `Engines/Compilation/Serialization/`.

**Tests (~25):** round-trip every format, HF tiny-model load + logits parity, `.pt` load end-to-end, TensorBoard output opens in real TB UI, FSDP sharded + reshard, CLI end-to-end.

**Benchmarks (~5):** `SafetensorsLoadBenchmarks`, `PtReaderBenchmarks`, `TensorBoardWriteBenchmarks`, `ShardedCheckpointBenchmarks`, `CrossFormatConvertBenchmarks`.

## 6. Acceptance gate

1. Round-trip every format.
2. Load public HF models (tiny-BERT/ViT-Tiny/tiny-Llama) + reference logits within 1e-4.
3. Sharded safetensors 10GB roundtrip.
4. `.pt` reader loads PyTorch state dict end-to-end.
5. TensorBoard output opens in real TB UI.
6. FSDP sharded save + reshard 4→2 ranks match.
7. CLI convert pt→safetensors→gguf works.
8. Safetensors load within 10% of Rust crate on Linux, wins on Windows.

## 7. Timeline — ~6 weeks

W1 SafetensorsReader (mmap, slices, metadata).
W2 SafetensorsWriter (consolidated + sharded + appendable).
W3 `.pt` / pickle parser + tensor recovery.
W4 HFStateDictMapper + 10 preset families + config.json.
W5 TensorBoard SummaryWriter (protobuf-direct).
W6 DistributedCheckpoint + CLI + benchmarks + PR.

## 8. Dependencies

- **Blocks**: #224 (PyTorch-compatible optimizer state-dict), #225 (model loading for AOT scenarios).
- **Blocked by**: #215 (FSDP for sharded checkpoint).

## 9. Risks

- **Pickle protocol subtleties** — malicious `.pt` files exist; ship read-only allowlist parser that refuses unknown opcodes.
- **HF naming convention drift** — pin each preset to a specific HF model-config version.
- **TensorBoard protobuf format stability** — test against current `tensorboard==2.15+`.

---

*Part of epic #209. Plan ratified 2026-04-19.*

**Comment by ooples** [comment]
## Final residual cleanup — PR #268

PR #268 closes the leftover acceptance gates on this issue:

- ✅ **CLI in solution**: `src/AiDotNet.Tensors.Cli` was orphaned from `AiDotNet.Tensors.slnx`; now wired up so CI builds it.
- ✅ **CLI smoke tests**: `CliConvertTests` drives `Program.Main` directly, round-trips safetensors → sharded → safetensors with bit-exact byte verification.
- ✅ **HF preset bijection sweep**: 130 cases across all 10 shipped presets. Surfaced and fixed two latent bugs in `HFStateDictMapper.MapReverse` (anchor-strip ordering corrupted `$1` backrefs; multi-capture rules like SD_UNET's `down[$1].attn[$2]` lost `$2`/`$3` on inversion).
- ✅ **HF synthetic-model parity test**: deterministic 2-layer MLP written under HF LLAMA naming, read back through the mapper, bit-exact weights + bit-exact forward outputs on single-file AND sharded paths.
- ✅ **5 benchmarks** per the implementation plan: `SafetensorsLoadBenchmarks`, `PtReaderBenchmarks`, `TensorBoardWriteBenchmarks`, `ShardedCheckpointBenchmarks`, `CrossFormatConvertBenchmarks`.

## FSDP reshard 4→2 — defers to #215

The acceptance item

> FSDP sharded checkpoint: save across 4 ranks, reshard to 2 ranks, weights match

is blocked on the distributed training stack (#215). FSDP itself doesn't exist yet on this side, so testing reshard semantics requires the ProcessGroup / DDP / FSDP infrastructure that #215 covers. The reshard test will land as part of #215's acceptance gate; the underlying `DistributedCheckpoint.Save/Load` API in `src/AiDotNet.Tensors/Serialization/Distributed/` is in place and unit-tested today, so #215 only needs to wire the FSDP-side ranks.
